Assets/Scripts/MovePlayer.cs

```diff
@@ -0,0 +1,237 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class MovePlayer : MonoBehaviour
+{
+    #region SerializeFields
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a slow movement (through sand, etc.).")]
+    private float slowMoveTime = .75f;
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a single movement.")]
+    private float moveTime = .5f;
+    [SerializeField]
+    [Tooltip("Threshold to stop movement coroutine and snap to position.")]
+    private float movementMargin = .05f;
+    [SerializeField]
+    [Tooltip("Point from which to raycast to check collision before moving.")]
+    private Transform collisionRaycastOrigin;
+    [SerializeField]
+    [Tooltip("Layer(s) the player collides with when moving.")]
+    private LayerMask collisionLayerMask;
+    [SerializeField]
+    [Tooltip("Layer that slows the player down when moving.")]
+    private LayerMask slowTerrainLayerMask;
+    [SerializeField]
+    [Tooltip("Tag of terrain that slows the player down when moving.")]
+    private string slowTerrainTag;
+    #endregion
+
+    /// <summary>
+    /// Defines directions in an enum for legible array reference
+    /// </summary>
+    private enum directions { Down, Up, Left, Right};
```

Comment:
Let's see if you can remove this and use a vector 2 movement direction plus built-in vector2.up,right, etc.